sci-libs/vtk/vtk-9.1.0.ebuild

```diff
@@ -0,0 +1,661 @@
+#	properly before building.
+# - replace usex by usev once we bump to EAPI 8
+
+PYTHON_COMPAT=( python3_{8..10} )
+WEBAPP_OPTIONAL=yes
+WEBAPP_MANUAL_SLOT=yes
+
+inherit check-reqs cmake cuda java-pkg-opt-2 python-single-r1 toolchain-funcs virtualx webapp
+
+# Short package version
+MY_PV="$(ver_cut 1-2)"
+
+DESCRIPTION="The Visualization Toolkit"
+HOMEPAGE="https://www.vtk.org/"
+SRC_URI="
+	https://www.vtk.org/files/release/${MY_PV}/VTK-${PV}.tar.gz
+	https://www.vtk.org/files/release/${MY_PV}/VTKData-${PV}.tar.gz
+	https://www.vtk.org/files/release/${MY_PV}/VTKDataFiles-${PV}.tar.gz
+	doc? ( https://www.vtk.org/files/release/${MY_PV}/vtkDocHtml-${PV}.tar.gz )
+	examples? (
+		https://www.vtk.org/files/release/${MY_PV}/VTKLargeData-${PV}.tar.gz
+		https://www.vtk.org/files/release/${MY_PV}/VTKLargeDataFiles-${PV}.tar.gz
+	)
+	test? (
+		https://www.vtk.org/files/release/${MY_PV}/VTKLargeData-${PV}.tar.gz
+		https://www.vtk.org/files/release/${MY_PV}/VTKLargeDataFiles-${PV}.tar.gz
+	)
+"
+S="${WORKDIR}/VTK-${PV}"
+
+LICENSE="BSD LGPL-2"
+SLOT="0/${MY_PV}"
+KEYWORDS="~amd64 ~arm ~arm64 ~x86 ~amd64-linux ~x86-linux"
+# TODO: Like to simplifiy these. Mostly the flags related to Groups, plus
+# maybe some flags related to Kits and a few other needed flags.
+IUSE="all-modules boost cuda debug doc examples ffmpeg gdal imaging java
+	mpi mysql odbc openmp postgres python qt5 +rendering tbb test +threads
+	tk video_cards_nvidia views web"
+
+RESTRICT="!test? ( test )"
+
+REQUIRED_USE="
+	all-modules? ( boost ffmpeg gdal imaging mysql odbc postgres qt5 rendering views )
+	cuda? ( video_cards_nvidia )
+	java? ( rendering )
+	python? ( ${PYTHON_REQUIRED_USE} )
+	qt5? ( rendering )
+	tk? ( rendering python )
+	web? ( python )
+"
+
+#	offscreen? ( media-libs/mesa[osmesa] )
+RDEPEND="
+	app-arch/lz4:=
+	app-arch/xz-utils
+	dev-db/sqlite:3
+	dev-libs/double-conversion:=
+	dev-libs/expat
+	dev-libs/icu:=
+	dev-libs/jsoncpp:=
+	dev-libs/libxml2:2
+	dev-libs/pugixml
+	media-libs/freetype
+	media-libs/libogg
+	media-libs/libpng:=
+	media-libs/libtheora
+	media-libs/tiff
+	<sci-libs/hdf5-1.12:=[mpi=]
+	sci-libs/netcdf:=[mpi=]
+	sys-libs/zlib
+	virtual/jpeg
+	all-modules? ( sci-geosciences/liblas[gdal] )
+	boost? ( dev-libs/boost:=[mpi?] )
+	cuda? ( dev-util/nvidia-cuda-toolkit:= )
+	ffmpeg? ( media-video/ffmpeg:= )
+	gdal? ( sci-libs/gdal:= )
+	java? ( >=virtual/jdk-1.8:* )
+	mpi? (
+		media-libs/glew:=
+		sys-cluster/openmpi[cxx,romio]
+		virtual/opengl
+	)
+	mysql? ( dev-db/mariadb-connector-c )
+	odbc? ( dev-db/unixODBC )
+	postgres? ( dev-db/postgresql:= )
+	python? ( ${PYTHON_DEPS} )
+	qt5? (
+		dev-qt/qtcore:5
+		dev-qt/qtopengl:5
+		dev-qt/qtquickcontrols2:5
+		dev-qt/qtsql:5
+		dev-qt/qtwidgets:5
+	)
+	rendering? (
+		media-libs/freeglut
+		media-libs/glew:=
+		media-libs/libsdl2
+		sci-libs/proj:=
+		virtual/opengl
+		x11-libs/gl2ps
+		x11-libs/libXcursor
+	)
+	tbb? ( <dev-cpp/tbb-2021.4.0:= )
```

Comment:
> Actually, vtk-9.1.0 seems to be already fixed to support new tbb-2021.4.0 (well, at least, as reported by the user provided my patches for it and 9.0.3).

>     * Core vtk libraries build successfully with latest tbb. If the vtkm third party module is getting build, the build fails with latest tbb

I had it updated at first, but later needed to re-add the restriction due to failures in the vtkm module.

Reply:
Well, I have patches (that I take from upstream today, to make 9.0.3 work) fto fix that.

Can you please navigate me, if there is any other non-resolved issue here?
If there is no - wouldn't you mind if I'll commit your ebuild (with my patches and some additions) and add you as the author?

Reply:
I'd much rather we wait until waebbl is happy and then commit it unkeyworded or masked for toralf or ago to do reverse dependency testing.

You can then make a PR with the changes you want to make in a new revision on top of that -- I'd prefer to review it separately.

Reply:
Actually, here is the whole difference:

```
--- vtk-9.1.0-r0.ebuild	2021-12-04 05:40:38.837523721 +0700
+++ vtk-9.1.0.ebuild	2021-12-04 05:41:27.377116532 +0700
@@ -107,7 +107,7 @@ RDEPEND="
 		x11-libs/gl2ps
 		x11-libs/libXcursor
 	)
-	tbb? ( <dev-cpp/tbb-2021.4.0:= )
+	tbb? ( dev-cpp/tbb:= )
 	tk? ( dev-lang/tk:= )
 	video_cards_nvidia? ( x11-drivers/nvidia-drivers[tools,static-libs] )
 	views? (
@@ -129,7 +129,7 @@ RDEPEND="
 DEPEND="
 	${RDEPEND}
 	dev-cpp/eigen
-	<dev-libs/pegtl-3
+	dev-libs/pegtl
 	dev-libs/utfcpp
 "
 BDEPEND="
@@ -145,6 +145,9 @@ BDEPEND="
 PATCHES=(
 	"${FILESDIR}"/${PN}-9.0.1-0001-fix-kepler-compute_arch-if-CUDA-toolkit-11-is-used.patch
 	"${FILESDIR}"/${PN}-8.2.0-freetype-2.10.3-provide-FT_CALLBACK_DEF.patch
+	"${FILESDIR}"/${PN}-9.0.3-cuda-11.5.0.patch
+	"${FILESDIR}"/${P}-pegtl-3.patch
+	"${FILESDIR}"/${P}-tbb-fix-for-bundled-vtkm.patch
 )
 
 DOCS=( CONTRIBUTING.md README.md )
```

And here is patches:
files/vtk-9.0.3-cuda-11.5.0.patch: https://gist.github.com/raw/be7ac35e2ff77c00896e9fb7885720c6
files/vtk-9.1.0-pegtl-3.patch: https://gist.github.com/raw/462381f875cca6fc96e3dd18aed18d93
files/vtk-9.1.0-tbb-fix-for-bundled-vtkm.patch: https://gist.github.com/raw/9f9cd7de552f5dcf779cc65416171658


And I will be okay even if you'll review them now and then @waebbl will take them to his PR and then agoand toralf will test it at whole (with support for pegtl-3 and tbb-2021).


Although, I still not finished building the 9.1.0, so maybe I'll add one or few patches a bit later.

Reply:
Also, maybe it would be wiser to specify dep atoms for `pegtl` and `tbb` as `<dev-libs/pegtl-4` and `<dev-cpp/tbb-2022:=` :man_shrugging:

Reply:
Please include links to where each of these patches are from. Several of them appear quite invasive.

Which are upstream? Which aren't upstream? Are there open bugs for these?

I'd prefer to not patch <9.1 at all.

As mentioned on IRC, several of these are also too big to go in the repository, especially combined. Will need to put the ones we want to take into a tarball (but we can handle that at the point of merging, it's easier to keep them in files/ for the purposes of review).

Reply:
> Please include links to where each of these patches are from

files/vtk-9.0.3-cuda-11.5.0.patch: As mentioned at the top of the file, the source is user who provided it.
I tested that it really builds with mentioned CUDA toolkit version.

files/vtk-9.1.0-pegtl-3.patch: Also, patch was made by @perestoronin by taking https://github.com/USBGuard/usbguard/pull/453 as hints.

files/vtk-9.1.0-tbb-fix-for-bundled-vtkm.patch:
this commits from vtk-m upstream repo:
https://gitlab.kitware.com/vtk/vtk-m/-/commit/97a2408fc
https://gitlab.kitware.com/vtk/vtk-m/-/commit/1eea0bee1
https://gitlab.kitware.com/vtk/vtk-m/-/commit/46b026882

Reply:
The CUDA patch and the pegtl one should both be submitted upstream or at least with a link to an upstream bug report. Gentoo is not maintaining a fork of VTK, so patches should go where they belong first.

I don't see the value in the TBB one right now given _very_ little works with newer TBB. Can you explain why it's necessary? And I'd still want waebbl's input there.

Reply:
> Can you explain why it's necessary

Not that it was "necessary" per se.

Just some hobby projects I'm using `vtk` in, already decided to move to `tbb-2021` and `pegtl-3` in the projects themselves (as both of them are already about a year old and have useful features).

And, as far as I get @perestoronin correctly, he is in the similar situation.

So, my intention was to contribute a bit of spare time after working day (well, actually, it took few days to test some combinations I thought of, and it anyway failed for Zorry 😿) to make it to work with  versions of deps that we're already using anyway (to avoid bundling on project side).


Anyway, I've got your point, and will try to make things right way when I'd be able to.

Reply:
The tbb patches haven't found their way into the 9.1.0 package, but are in the current master, and thus will get into the next release of vtk. Given that, and what @thesamesam said, only very few packages actually support tbb-2021, I think there's no need to rush with these patches. But I'm not against adding them. If we decide to add them, I think, it's more advisable to split the patch to make it more manageable.

Haven't yet tested the pegtl-3 patch. It looks good to me, but I'd prefer if we open an issue upstream and provide a patch for it, before we add it to the tree. People who urgently need this patch for their own projects could be pointed to use `/etc/portage/patches` for this purpose.

Does the cuda patch also work for <cuda-11.5? I can't test with cuda-11.5, as it removed support for my sm_35 based GPU arch. Does it work with both gcc-10 and gcc-11? Given the lot of issues we lately had with cuda in vtk, I'm thinking about masking the cuda use flag for the time being, until we have figured out how to solve it in a stable manner. Building with cuda support has some meaningful hardware requirements and build times, so I'd rather be more conservative when it comes to patching these. Additionally this is another patch, where we should go upstream first and report it and try to help them figuring and fixing it.